### DISCUSSION TITLE
Using the athena code for non-ideal MHD simulations

### DISCUSSION THREAD

**Discussion by rahulgaur104** [PrincetonUniversity/athena#401] [Q&A] [answered]
Hi everyone,

I am a fifth-year graduate student at the University of Maryland, College Park. I work on fusion plasma turbulence. I want to simulate a plasma equilibrium for a rapidly(sound Mach number ~ 5) rotating plasma in cylindrical geometry using the non-ideal(resistive and viscous) MHD equations. Is it possible to solve non-ideal MHD equations using this repo of athena? If so, could you point me towards an example or tell me how to turn on resistivity and viscosity? All the problem generator files I have looked at either work with ideal MHD or use resistivity(resist.cpp) only during the initial setup.

Best regards,
Rahul Gaur

**Comment by c-white** [comment, marked as answer]
Non-ideal physics is described under [Diffusion Processes](https://github.com/PrincetonUniversity/athena/wiki/Diffusion-Processes). In order to turn on constant momentum, thermal, ohmic, and/or ambipolar diffusivities (units of code length squared per code time), just set the variables `nu_iso`, `kappa_iso`, `eta_ohm`, and/or `eta_ad` in the `<problem>` section of the input file. There are no configuration changes, nor does the problem generator need to be modified. Note that there are suggestive variables `nu_aniso`, `kappa_aniso`, and `eta_hall` that the code will read and process, but I do not believe anything will be done with them.

Non-constant diffusivities should also be possible, but that requires writing and enrolling certain functions in the problem generator.

I don't know if there are any test problems in the code for non-ideal physics + curvilinear coordinates. These should work together, but I recommend running a simple problem to make sure the code is doing what you think it's doing.

By default, the code will take explicit steps with diffusive terms. These might become much smaller than the sound crossing times, slowing down the calculation. If that is an issue, you should look into turning on [super time stepping](https://github.com/PrincetonUniversity/athena/wiki/Super-Time-Stepping).

**Comment by rahulgaur104** [comment]
Thanks so much, Chris! I'll try to run a simple test case and share my input file with you by the end of this week.

**Comment by rahulgaur104** [comment]
Hi Chris and Kyle(@c-white, @felker),

Here are all the details of the simple test problem that I am trying to solve:
## Problem definition

I am trying to solve an axisymmetric, axially symmetric MHD problem in cylindrical coordinates 
<a href="https://www.codecogs.com/eqnedit.php?latex=\large&space;\partial()/\partial&space;\theta&space;=&space;0,&space;\partial()/\partial&space;z&space;=&space;0" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\large&space;\partial()/\partial&space;\theta&space;=&space;0,&space;\partial()/\partial&space;z&space;=&space;0" title="\large \partial()/\partial \theta = 0, \partial()/\partial z = 0" /></a>

I have turned on resistivity and viscosity. The system is isothermal. The initial conditions are as follows:
<img src="https://latex.codecogs.com/gif.latex?\large&space;\begin{equation}&space;\begin{gathered}&space;{\color{black}u_r&space;=0,&space;u_z&space;=&space;0,&space;u_{\theta}&space;=&space;c_s&space;(r-a)(r-b)}\\&space;{\color{black}B_r&space;=0,&space;B_z&space;=&space;1,&space;B_{\theta}&space;=&space;0}&space;\end{gathered}&space;\end{equation}" title="\large \begin{equation} \begin{gathered} {\color{black}u_r =0, u_z = 0, u_{\theta} = c_s (r-a)(r-b)}\\ {\color{black}B_r =0, B_z = 1, B_{\theta} = 0} \end{gathered} \end{equation}" />

where a=0.15 and b=0.5 as you can see the attached problem file. Assuming radial force balance during the initialization, we must satisfy
<a href="https://www.codecogs.com/eqnedit.php?latex=\large&space;\begin{equation}&space;{\color{black}n&space;m&space;\dfrac{u_{\theta}^2}{r}&space;=&space;T&space;\dfrac{dn}{dr}}&space;\end{equation}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\large&space;\begin{equation}&space;{\color{black}n&space;m&space;\dfrac{u_{\theta}^2}{r}&space;=&space;T&space;\dfrac{dn}{dr}}&space;\end{equation}" title="\large \begin{equation} {\color{black}n m \dfrac{u_{\theta}^2}{r} = T \dfrac{dn}{dr}} \end{equation}" /></a>

I have assumed m = 1. Integrating the radial force balance equation, we obtain the initial density 
<img src="https://latex.codecogs.com/gif.latex?n&space;=&space;n_0&space;\exp\left(\int_{a}^r&space;\frac{u_{\theta}^2}{r}\right)" title="n = n_0 \exp\left(\int_{a}^r \frac{u_{\theta}^2}{r}\right)" />

which has been calculated exactly and is given in the attached pgen file. I have assumed n_0 = 1. The user-defined boundary conditions are as follows:
<img src="https://latex.codecogs.com/gif.latex?\large&space;\begin{equation}&space;\begin{gathered}&space;{\color{black}B_{\theta}(r&space;=&space;a,b)&space;=&space;0,&space;\frac{\partial&space;B_r}{\partial&space;r}(r&space;=&space;a,&space;b)&space;=&space;0,&space;\frac{\partial&space;B_z}{\partial&space;r}(r&space;=&space;a,&space;b)&space;=&space;0}\\&space;{\color{black}u_{r}(r&space;=&space;a,b)&space;=&space;0,&space;u_{\theta}(r&space;=&space;a,b)&space;=&space;0,&space;\frac{\partial&space;u_z}{\partial&space;r}&space;=&space;0}\\&space;{\color{black}&space;\frac{\partial&space;B_r}{\partial&space;z}(z&space;=&space;\pm&space;z_0)&space;=&space;0,&space;\frac{\partial&space;B_{\theta}}{\partial&space;z}(z&space;=&space;\pm&space;z_0),&space;\frac{\partial&space;B_z}{\partial&space;z}(z&space;=&space;\pm&space;z_0)&space;=&space;0}\\&space;{\color{black}\frac{\partial&space;u_{r}}{\partial&space;z}(z&space;=&space;\pm&space;z_0)&space;=&space;0,&space;\frac{\partial&space;u_{\theta}}{\partial&space;z}(z&space;=&space;\pm&space;z_0)&space;=&space;0,&space;u_z(z&space;=&space;\pm&space;z_0)}&space;\end{equation}&space;\end{gathered}" title="\large \begin{equation} \begin{gathered} {\color{black}B_{\theta}(r = a,b) = 0, \frac{\partial B_r}{\partial r}(r = a, b) = 0, \frac{\partial B_z}{\partial r}(r = a, b) = 0}\\ {\color{black}u_{r}(r = a,b) = 0, u_{\theta}(r = a,b) = 0, \frac{\partial u_z}{\partial r} = 0}\\ {\color{black} \frac{\partial B_r}{\partial z}(z = \pm z_0) = 0, \frac{\partial B_{\theta}}{\partial z}(z = \pm z_0), \frac{\partial B_z}{\partial z}(z = \pm z_0) = 0}\\ {\color{black}\frac{\partial u_{r}}{\partial z}(z = \pm z_0) = 0, \frac{\partial u_{\theta}}{\partial z}(z = \pm z_0) = 0, u_z(z = \pm z_0)} \end{equation} \end{gathered}" />

and periodic boundary conditions in theta. z_0=0.5 is the position of the Z boundary. 

## Outputs
Upon configuring the custom problem, compiling the code and running with the attached input file I get the following results. 
* magnitude of the magnetic field becomes NaN
<a href="https://ibb.co/zFyf9m4"><img src="https://i.ibb.co/zFyf9m4/Bmag-movie0000.png" alt="Bmag-movie0000" border="0"></a> <a href="https://ibb.co/4P9WCt6"><img src="https://i.ibb.co/4P9WCt6/Bmag-movie0001.png" alt="Bmag-movie0001" border="0"></a> <a href="https://ibb.co/KL5WGXg"><img src="https://i.ibb.co/KL5WGXg/Bmag-movie0002.png" alt="Bmag-movie0002" border="0"></a> <a href="https://ibb.co/CVVcrq9"><img src="https://i.ibb.co/CVVcrq9/Bmag-movie0003.png" alt="Bmag-movie0003" border="0"></a>

* the azimuthal flow speed vel2 becomes NaN
<a href="https://ibb.co/GRXsM0Y"><img src="https://i.ibb.co/GRXsM0Y/utheta-movie0000.png" alt="utheta-movie0000" border="0"></a> <a href="https://ibb.co/sqsw2rP"><img src="https://i.ibb.co/sqsw2rP/utheta-movie0001.png" alt="utheta-movie0001" border="0"></a> <a href="https://ibb.co/yWqmg8y"><img src="https://i.ibb.co/yWqmg8y/utheta-movie0002.png" alt="utheta-movie0002" border="0"></a> <a href="https://ibb.co/b5SCnwD"><img src="https://i.ibb.co/b5SCnwD/utheta-movie0003.png" alt="utheta-movie0003" border="0"></a>


* density goes to 0
<a href="https://ibb.co/KKGzwkC"><img src="https://i.ibb.co/KKGzwkC/rho-movie0000.png" alt="rho-movie0000" border="0"></a> <a href="https://ibb.co/d4jCgJt"><img src="https://i.ibb.co/d4jCgJt/rho-movie0001.png" alt="rho-movie0001" border="0"></a> <a href="https://ibb.co/3c3bBmz"><img src="https://i.ibb.co/3c3bBmz/rho-movie0002.png" alt="rho-movie0002" border="0"></a> <a href="https://ibb.co/3FpVXgH"><img src="https://i.ibb.co/3FpVXgH/rho-movie0003.png" alt="rho-movie0003" border="0"></a>

## Related queries
The density should satisfy Neumann BC according to the problem file but it's 0 at r = a, b. Similarly, I want B_z =B_0 everywhere, even at the boundaries. I don't know how to do that. Finally, there could be a problem with my understanding of the normalizations. It would be helpful if you could explain how normalizations work in Athena++.

I would be immensely grateful for any help/suggestion. I have attached the problem generator and input file. 
[files.zip](https://github.com/PrincetonUniversity/athena/files/7598281/files.zip)

**Reply by c-white** [reply]
First, some things I'm noticing looking at the files:
- `x1rat`, `x2rat`, and `x3rat` are not limits on aspect ratio; rather they force adjacent cell widths in that direction to have that ratio. So by having these not be 1, the grid is not uniformly spaced in the given direction. It's unusual to want non-uniform spacing in z, and I'm certain you don't want it in azimuth.
- The primitives contain velocities (indexed with `IVX`, `IVY`, and `IVZ`), not momenta (`IM1`, `IM2`, `IM3`). Your code works, and it turns out these two sets of indices are the same (internally they happen to be 1, 2, and 3), but I just wanted to make sure there was no confusion.
- Between `b.x1f(k,j,iu+i+1) = b.x1f(k,j,iu+1);` (line 262) and `b.x3f(ku+k+1,j,i) = b.x3f(ku,j,i);` (line 382), it seems like these are two different approaches to the 0-slope boundary condition.
- How is this being configured/compiled? The comment in the input file (and this is just a comment ignored by the code) says `-b -sts --prob=mirror_simpler`, but surely you also want `--coord cylindrical --eos isothermal -hdf5` as well.

That said, I think I could reproduce the problem. It persisted after turning off diffusion and sts, but then disappeared after setting all cell width ratios to 1. Can you confirm that the problem works if the ratios are set to 1?

Still, `x1rat = 1.05` is a perfectly valid request and should work. Very preliminarily, it looks like with that set, reconstruction in the radial direction near the outer boundary produces a 0 in the density, which it shouldn't do.

**Reply by c-white** [reply]
Also, I'm not quite sure what you mean by normalizations.

**Reply by c-white** [reply]
I think I found it! The user-defined boundary functions are declaring their inputs in the wrong order. The last arguments should be `int il, int iu, int jl, int ju, int kl, int ku, int ngh`. Right now, these values are being scrambled (but the rest of the function definition is correct).

**Reply by rahulgaur104** [reply]
Hi Chris(@c-white),

Thanks for your help!

Yes, I want a unity aspect ratio. I'll remove the variables `xrat1`, `xrat2`, `xrat3` from the input file.

I'll edit the problem generator file to set the velocities instead of momenta after I get the code running.

I have changed the line 382 to `b.x3f(ku+k+1,j,i) = b.x3f(ku+1,j,i);` I hope that's the correct way to impose Neumann condition.

This is my cofigure command:

`python3 configure.py --prob=mirror_simpler --coord=cylindrical --eos=isothermal -b -sts -mpi-hdf5 --hdf5_path=/usr/local/hdf5/gcc/openmpi-4.1.0/1.10.6 --lib_path=/usr/local/hdf5/gcc/openmpi-4.1.0/1.10.6/lib64`

By normalizations, I want to ask if variables like temperature, density etc are normalized w.r.t to some value or are they specified in some standard unit system(SI, CGS)? I have seen sound speed c_s = 1. That makes me think that the velocities are normalized w.r.t the sound speed. Are there similar normalizations for the pressure, magnetic field?

I changed the order of declaration and the code seems to work without NaNs. I'll attach the correct test problem and input files after I have checked my results.

**Reply by rahulgaur104** [reply]
Perhaps I should start a new discussion inquiring about normalizations/units.

**Reply by felker** [reply]
See this set of slides: https://gandalfcode.github.io/gandalf-school/Units.pdf
Athena++'s use of arbitrary dimensionless units currently follows the “Correct, but more difficult than it needs to be” approach on the second slide, but it would be great to add the recommended functionality for "Use dimensionless units where the scale factors are calculated by the code automatically". I believe Enzo also does this. 

also: https://github.com/PrincetonUniversity/athena/wiki/Diffusion-Processes#units

**Reply by rahulgaur104** [reply]
Hi Kyle,

Thanks for the reply! I looked at that presentation and the link to the Diffusion-Processes page. I am still confused as to how the normalizations work in ATHENA++. If I assume that ATHENA++ is using the same procedure as GANDALF for the hydro package, then the following scaling relation must be true

<a href="https://www.codecogs.com/eqnedit.php?latex=\LARGE&space;\frac{k_B&space;\theta_0}{\bar{m}\,&space;V_0^2}&space;=&space;1" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\LARGE&space;\frac{k_B&space;\theta_0}{\bar{m}\,&space;V_0^2}&space;=&space;1" title="\LARGE \frac{k_B \theta_0}{\bar{m}\, V_0^2} = 1" /></a>

which comes from the definition of the sound speed and the need to remove the factor k_B/m. But there many more normalizing variables such as the density(n_0), magnetic field(B_0), length(L_0) and time(T_0).  We would need as many scaling relations as the normalizing variables. What are the rest of the scaling relations?

What about magnetic field normalization? Do we use the Alfven speed?

<a href="https://www.codecogs.com/eqnedit.php?latex=\LARGE&space;v_A&space;=&space;\sqrt{\frac{B^2}{\mu_0&space;\bar{m}&space;n}}\\&space;v_A^{'}&space;=&space;\underbrace{\sqrt{\frac{B_0^2}{\mu_0&space;\bar{m}&space;n_0&space;V_0^2}}}_{1}&space;\sqrt{\frac{{B^{'}}^2}{n^{'}}}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\LARGE&space;v_A&space;=&space;\sqrt{\frac{B^2}{\mu_0&space;\bar{m}&space;n}}\\&space;v_A^{'}&space;=&space;\underbrace{\sqrt{\frac{B_0^2}{\mu_0&space;\bar{m}&space;n_0&space;V_0^2}}}_{1}&space;\sqrt{\frac{{B^{'}}^2}{n^{'}}}" title="\LARGE v_A = \sqrt{\frac{B^2}{\mu_0 \bar{m} n}}\\ v_A^{'} = \underbrace{\sqrt{\frac{B_0^2}{\mu_0 \bar{m} n_0 V_0^2}}}_{1} \sqrt{\frac{{B^{'}}^2}{n^{'}}}" /></a>

I would appreciate any help. Thanks!

**Reply by felker** [reply]
The hydro and MHD scaling constraints you listed seem correct to me, since we indeed calculate the Alfven speed in dimensionless internal code units as v_a = |B|/sqrt(dens). 

I dont have any experience converting the output to SI units, however.

**Reply by c-white** [reply]
I'm not sure I understand all the notation above, so I'll try to explain normalizations from scratch.

The first thing to remember is that Athena++ is meant for a broader range of astrophysics problems than most other codes (e.g. Gandalf, Enzo, Arepo), so there is no natural physical scale to default to. You might be simulating a cubic millimeter of plasma or a galaxy.

For hydrodynamics, to convert between physical units and code units, you need three independent scales. It's probably easiest to choose length [L], time [t], and density [&rho;]. So for example, if [L] = 1 km, then a code simulation running from x = 0 to x = 10 covers a length of 10 km. In any case we have the following dictionary:

| Quantity | Code value | Physical value |
| - | - | - |
| Position | x | x [L] |
| Time | t | t [t] |
| Density | &rho; | &rho; [&rho;] |
| Velocity | v | v [L] [t]<sup>-1</sup> |
| Pressure | p | p [&rho;] [L]<sup>2</sup> [t]<sup>-2</sup> |
| Energy density | u | u [&rho;] [L]<sup>2</sup> [t]<sup>-2</sup> |
| Momentum density | m | m [&rho;] [L] [t]<sup>-1</sup> |
| Mass | M | M [&rho;] [L]<sup>3</sup> |
| Diffusivity | &eta; | &eta; [L]<sup>2</sup> [t]<sup>-1</sup> |

There is no temperature or entropy, because these quantities do not enter into the equations of fluid dynamics. (If you have thermal diffusivity or radiation, then you need to know that the code treats p/&rho; as "temperature." That is, k<sub>B</sub> / (&mu; m<sub>p</sub>) = 1 in code units.)

For the magnetic field, Athena++ is based on Lorentz-Heaviside units (rather than CGS, which makes the MHD equations slightly more complicated, or SI, which makes a mess of E&M entirely). For hydro, this distinction doesn't matter, since the above relations are true in all decent unit systems. To convert magnetic fields back to CGS, the dictionary includes:

| Quantity | Code value | Physical value in CGS |
| - | - | - |
| Magnetic field squared | B<sup>2</sup> | 4 &pi; B<sup>2</sup> [&rho;] [L]<sup>2</sup> [t]<sup>-2</sup> |
| Magnetic field | B | (4 &pi;)<sup>1/2</sup> B [&rho;]<sup>1/2</sup> [L] [t]<sup>-1</sup> |

As a check, if [L] = 1 m, [t] = 1 s, and [&rho;] = 1 kg/m<sup>3</sup>, then 1 code unit of B corresponds to about 11.2 gauss.

Wavespeeds don't come into this, except accidentally. Sound waves in Newtonian physics move at c<sub>s</sub> = (&Gamma; p / &rho;)<sup>1/2</sup>, both in LH/CGS/SI and in Athena++, so I suppose if p = 1 and &rho; = 1 in code units, **and** if &Gamma; = 1 (i.e., you are using an isothermal EOS), then c<sub>s</sub> = 1 in code units. Similarly, Alfven waves move at v<sub>A</sub> = (B^2 / &rho;)<sup>1/2</sup> in LH and Athena++ (but not CGS or SI), so for B = 1 and &rho; = 1 in code units, v<sub>A</sub> = 1 in code units. But the above dictionary is derived based on dimensional analysis with no unnecessary constants; it just so happens that the formulas for some speeds also have no extra constants.

**Reply by rahulgaur104** [reply]
Hi Chris,

Thanks for such a detailed reply!

Let me convert all the variables from the code output to their dimensional values for a given set of values and you can tell me if I understand it correctly. At a given region in the magnetic mirror, say I have the following values in code units

Simulation run time  = 1
Length of the cylinder = 1
Density of the plasma = 1
Pressure = 10
Magnetic field = 1
Sound speed(gamma = 1) = sqrt(10)
Alfven speed = 1

Since, we can choose [L], [t] and [ρ] freely, I choose [L] = 1m, [t] = 1s, [ρ] = 0.001 Kg/m^3. This would mean that the dimensional values are

Simulation run time  = 1s
Length of the cylinder = 1m
Density of the plasma = 0.001 Kg/m^3
Pressure = 0.01 Pa
Magnetic field = 0.35 G  = 3.5E-5 T
Sound speed(gamma = 1) = sqrt(10) m/s= 3.16 m/s
Afven speed =  3.5E-5/sqrt(4pi rho) = 3.1E-4 m/s

Is this conversion correct?

Thanks!

**Reply by c-white** [reply]
Everything is right except the Alfven speed. All velocities are converted with the same factor, here [L]/[t] = 1 m/s, so 1 becomes 1 m/s. Alternatively, calculate it in CGS: 0.35 gauss / (4 * pi * 1e-6 g/cc) = 100 cm/s. You can calculate it in SI using 3.5e-5 T and 0.001 kg/m^3, but the formula changes and probably has &epsilon;<sub>0</sub> somewhere in it.

**Reply by rahulgaur104** [reply]
Got it. Thanks, Chris!
